### PR TITLE
[WIP] TRT-2741: Add cumulative summary tables for matview refresh optimization

### PR DESCRIPTION
**Pull request by mstaeble**
## Summary

Add foundation tables for computing arbitrary date range aggregates
without expensive matview refreshes. This is Phase 1 (additive only,
nothing consumes these tables yet).

### New tables
- `test_cumulative_summaries`: running totals from `test_daily_summaries`, keyed by immutable fields (no variant_combination_id). Entities carry forward on gap days via FULL OUTER JOIN.
- `variant_cumulative_summaries`: groups `test_cumulative_summaries` by `variant_combination_id`. Updated daily by grouping one day's cumulative summaries. Variant changes detected via `vcid_mappings` trigger scoped rebuilds.
- `vcid_mappings`: tracks prow_job variant_combination_id mapping for change detection.

### New commands
- `sippy backfill --table TABLE --start-date DATE [--end-date DATE]`: targeted backfill of a specific table for a date range
- `scripts/backfill-summaries.sh`: shell script for gradually backfilling via pods

### Other changes
- Daily summary lookback extended from 14 to 90 days
- Releases queried from `release_definitions` (newest first) for better parallel worker distribution
- Session settings: `idle_in_transaction_session_timeout=60s`, `random_page_cost=1.1`
- `dailysummary.Refresh` returns earliest changed date for downstream consumers
- Separate `Refresh` and `Backfill` entry points in all summary packages

### Notable behavior changes
- `--rebuild-daily-summaries` flag removed from `sippy refresh` to prevent accidental data loss. Use `sippy backfill` instead.
- `scripts/updatesuites.go` previously used `Rebuild: true` to truncate and rebuild daily summaries after renaming tests. It now does an incremental refresh only. If a full rebuild is needed after running updatesuites, use `sippy backfill --table daily-summaries`.

## Test plan
- [ ] Unit tests pass (`go test ./pkg/...`)
- [ ] Build succeeds (`go build ./pkg/...`)
- [ ] Deploy to staging and run backfill script for cumulative summaries
- [ ] Verify cumulative summary correctness against direct SUM of daily summaries
- [ ] Verify incremental refresh processes only new dates

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a **`backfill`** CLI command for filling summary tables over a UTC date range.
  * Added cumulative and variant cumulative refresh/backfill support, including table-level dispatch.
  * Added a Kubernetes batch script to run backfills in short-lived pods.
* **Bug Fixes**
  * Refresh operations now **fail fast** and surface errors instead of continuing silently.
  * Materialized view refresh now reports refresh failures.
* **Chores**
  * Expanded partitioned tables and updated DB runtime settings; refreshed seeding to use the new backfill/refresh flow.
* **Tests**
  * Added unit tests covering cumulative refresh/backfill date-window edge cases.

### CODE REVIEW

**Comment by mstaeble** [comment]
@coderabbitai pause

File: cmd/sippy/backfill.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			endDate := civil.DateOf(time.Now().UTC())
+			if f.EndDate != "" {
+				endDate, err = civil.ParseDate(f.EndDate)
+				if err != nil {
+					return fmt.Errorf("invalid --end-date %q: %w", f.EndDate, err)
+				}
+			}
+
+			if startDate.After(endDate) {
+				return fmt.Errorf("--start-date (%s) is after --end-date (%s)", f.StartDate, f.EndDate)
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply cmd/sippy/backfill.go]: Fixed. Switched to the parsed startDate/endDate values.

File: pkg/db/cumulativesummary/variant_cumulative_summary.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+func (s *pgVariantStore) UpdateVCIDMapping() error {
+	if err := s.dbc.DB.Exec("TRUNCATE vcid_mappings").Error; err != nil {
+		return err
+	}
+	return s.dbc.DB.Exec(`
+		INSERT INTO vcid_mappings (prow_job_id, variant_combination_id)
+		SELECT id, variant_combination_id FROM prow_jobs
+	`).Error
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/db/cumulativesummary/variant_cumulative_summary.go:243]: Wrapped both statements in a dbc.DB.Transaction().

File: scripts/backfill-summaries.sh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        exit 1
+    fi
+    echo "Using image: $IMAGE"
+fi
+
+today=$(date -u +%Y-%m-%d)
+completed=0
+failed=0
+
+for ((offset=DAYS; offset>0; offset-=BATCH)); do
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply scripts/backfill-summaries.sh:85]: Added validation that DAYS, PAUSE, and BATCH are positive integers before the loop starts.

**Comment by mstaeble** [comment]
@coderabbitai resume

**Comment by mstaeble** [comment]
@coderabbitai full review

File: pkg/db/cumulativesummary/cumulative_summary_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	maxDailySummaryDate := civil.Date{Year: 2026, Month: 7, Day: 7}
+	earliestChanged := civil.Date{Year: 2026, Month: 7, Day: 7}
+	store := &fakeStore{maxCumulativeSummary: &maxCum, maxDailySummaryDate: &maxDailySummaryDate}
+
+	_, err := doRefresh(store, earliestChanged)
+
+	require.NoError(t, err)
+	assert.Len(t, store.updatedDates, 1)
+	assert.Equal(t, civil.Date{Year: 2026, Month: 7, Day: 7}, store.updatedDates[0])
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/db/cumulativesummary/cumulative_summary_test.go:128]: Valid concern. Will inject today into doRefresh so tests use a fixed reference date.

File: pkg/db/cumulativesummary/date_range.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package cumulativesummary
+
+import (
+	"cloud.google.com/go/civil"
+	log "github.com/sirupsen/logrus"
+)
+
+const maxAutoFillDays = 14
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/db/cumulativesummary/date_range.go:8]: This is intentional. Daily summary refresh is also capped at 14 days (defaultLookbackDays = 14). Both layers use the same 14-day cap during normal refresh. For anything older, the sippy backfill command should be used.

File: pkg/db/cumulativesummary/variant_cumulative_summary.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+		return err
+	}
+
+	if err := store.UpdateVCIDMapping(); err != nil {
+		return fmt.Errorf("updating VCID mapping: %w", err)
+	}
+
+	log.WithField("elapsed", time.Since(loadStart)).Info("variant cumulative summary refresh complete")
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/db/cumulativesummary/variant_cumulative_summary.go:90]: Agreed, will add tests for the variant refresh orchestration.

File: pkg/db/cumulativesummary/variant_cumulative_summary.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+			SUM(ps.prefix_sum_successes)::bigint, SUM(ps.prefix_sum_failures)::bigint,
+			SUM(ps.prefix_sum_flakes)::bigint, SUM(ps.prefix_sum_runs)::bigint
+		FROM test_cumulative_summaries ps
+		JOIN prow_jobs pj ON ps.prow_job_id = pj.id
+		JOIN affected_entities ae ON ps.test_id = ae.test_id
+			AND ps.suite_id = ae.suite_id AND ps.release = ae.release
+		WHERE pj.variant_combination_id IS NOT NULL
+		  AND ps.summary_date >= CURRENT_DATE - 90
+		GROUP BY ps.summary_date, ps.test_id, ps.suite_id, pj.variant_combination_id, ps.release
+	`, changedProwJobIDs).Error
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/db/cumulativesummary/variant_cumulative_summary.go:230]: Good catch. The DELETE scope is intentionally broader to clean up stale variant groupings across all dates. Rows older than 90 days are not consumed by any query path, so losing them is acceptable. If a full historical rebuild is needed, the sippy backfill command covers it.

File: pkg/flags/postgres_benchmarking_test.go [hunk 8]
```diff
@@ -1231,7 +1230,7 @@ func Test_BenchmarkRefreshData(t *testing.T) {
 	r := runBenchmarkCase(t, dbc, benchmarkCase{
 		name: "RefreshData",
 		fn: func(dbc *db.DB) error {
-			sippyserver.RefreshData(dbc, nil, false, dailysummary.Options{})
+			sippyserver.RefreshData(dbc, nil, sippyserver.RefreshOptions{})
 			return nil
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/flags/postgres_benchmarking_test.go]: Good catch, will update the benchmark to handle the error.

File: scripts/backfill-summaries.sh [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    batch_end_offset=$((offset - BATCH))
+    if (( batch_end_offset < 0 )); then
+        batch_end_offset=0
+    fi
+
+    start_date=$(date -u -v-${offset}d +%Y-%m-%d 2>/dev/null || date -u -d "$today - ${offset} days" +%Y-%m-%d)
+    end_date=$(date -u -v-${batch_end_offset}d +%Y-%m-%d 2>/dev/null || date -u -d "$today - ${batch_end_offset} days" +%Y-%m-%d)
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply scripts/backfill-summaries.sh:92]: Good catch, the range is inclusive on both ends. Will fix the date arithmetic.

File: scripts/backfill-summaries.sh [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            }
+        }" 2>&1 | { grep -iE 'summar|error|elapsed|complete' || true; }
+    exit_code=${PIPESTATUS[0]}
+    if [[ $exit_code -eq 0 ]]; then
+        completed=$((completed + 1))
+        echo "  Done."
+    else
+        failed=$((failed + 1))
+        echo "  FAILED (exit $exit_code)" >&2
+    fi
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply scripts/backfill-summaries.sh:133]: Valid. Will add || true to the oc run command so the script can handle failures itself.

File: scripts/updatesuites.go [hunk 11]
```diff
@@ -154,7 +153,7 @@ func testNameWithoutSuite(dbc *gorm.DB) error {
 	// NOTE: does not update timestamps to invalidate cached matview data; not clear if the use case for this script requires that.
 	sippyserver.RefreshData(&db.DB{
 		DB: dbc,
-	}, nil, false, dailysummary.Options{Rebuild: true})
+	}, nil, sippyserver.RefreshOptions{})
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply scripts/updatesuites.go]: Good catch, will handle the returned error.

File: pkg/sippyserver/server.go [hunk 12]
```diff
@@ -387,16 +395,57 @@ func recordMatviewRefreshTime(cacheClient cache.Cache, matView string, tmpLog *l
...
+	if err := cumulativesummary.RefreshVariantCumulativeSummaries(dbc, earliestCumulativeChanged); err != nil {
+		return fmt.Errorf("failed to refresh variant cumulative summaries: %w", err)
 	}
-	refreshMaterializedViews(dbc, cacheClient, refreshMatviewsOnlyIfEmpty)
+	cumulativeSummaryRefreshMetric.Observe(float64(time.Since(cumulativeStart).Milliseconds()))
+
+	refreshMaterializedViews(dbc, cacheClient, opts.RefreshOnlyIfEmpty)
 	log.Info("Refresh complete")
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/sippyserver/server.go:434]: Good catch, will update the benchmark to handle the error.

File: pkg/db/dailysummary/dailysummary.go [hunk 13]
```diff
@@ -155,78 +157,41 @@ func aggregateReleases(store summaryStore, releases []string, startDate, endDate
...
-			return *maxSummary, nil
+			return *maxSummary
 		}
-		return yesterday, nil
+		return yesterday
 	}
-
-	return now.AddDate(0, 0, -defaultLookbackDays), nil
+	return today.AddDays(-defaultLookbackDays)
 }
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/db/dailysummary/dailysummary.go:209]: This is intentionally 14. We don't want the first refresh to scan 90 days of prow_job_run_tests across all releases. The sippy backfill command should be used to populate older data. Both daily and cumulative summaries use matching 14-day caps during normal refresh.

**Comment by mstaeble** [comment]
Superseded by a new PR with partitioned tables.
